src/components/Editor/SymbolModal.js

```diff
@@ -0,0 +1,291 @@
+    });
+  }
+
+  componentDidUpdate(prevProps: any, prevState: any) {
+    if (this.refs.resultList && this.refs.resultList.refs) {
+      scrollList(this.refs.resultList.refs, this.state.selectedResultIndex);
+    }
+  }
+
+  onEscape(e: SyntheticEvent) {
+    e.stopPropagation();
+    e.preventDefault();
+    this.props.toggleActiveSearch();
+  }
+
+  onChange(e: SyntheticEvent) {
+    const { selectedSource, setSymbolSearchQuery } = this.props;
+    if (!selectedSource || !selectedSource.get("text")) {
+      return;
+    }
+
+    setSymbolSearchQuery(e.target.value);
+
+    return this.updateResults(e.target.value);
+  }
+
+  updateResults(query: string) {
+    const {
+      selectedSource,
+      updateSearchResults,
+      updateSymbolSearchResults,
+      selectedSymbolType,
+      symbols
+    } = this.props;
+
+    console.log(query, selectedSource);
+
+    if (query == "" || !selectedSource) {
+      return;
+    }
+
+    const symbolSearchResults = filter(symbols[selectedSymbolType], query, {
+      key: "value"
+    });
+
+    updateSearchResults({ count: symbolSearchResults.length });
+    updateSymbolSearchResults(symbolSearchResults);
+  }
+
+  traverseResults(rev: boolean) {
+    const { symbolSearchResults, selectedResultIndex } = this.state;
+    const searchResults = symbolSearchResults;
+    const resultCount = searchResults.length;
+
+    if (rev) {
+      let nextResultIndex = Math.max(0, selectedResultIndex - 1);
+
+      if (selectedResultIndex === 0) {
+        nextResultIndex = resultCount - 1;
+      }
+      this.setState({ selectedResultIndex: nextResultIndex });
+      this.onSelectResultItem(searchResults[nextResultIndex]);
+    } else {
+      let nextResultIndex = Math.min(resultCount - 1, selectedResultIndex + 1);
+
+      if (selectedResultIndex === resultCount - 1) {
+        nextResultIndex = 0;
+      }
+      this.setState({ selectedResultIndex: nextResultIndex });
+      this.onSelectResultItem(searchResults[nextResultIndex]);
+    }
+  }
+
+  onKeyDown(e: SyntheticKeyboardEvent) {
+    const { enabled, query } = this.props;
+    const { symbolSearchResults } = this.state;
+    if (!enabled || query == "") {
+      return;
+    }
+
+    const searchResults = symbolSearchResults;
+
+    if (e.key === "ArrowUp") {
+      this.traverseResults(true);
+      e.preventDefault();
+    } else if (e.key === "ArrowDown") {
+      this.traverseResults(false);
+      e.preventDefault();
+    } else if (e.key === "Enter") {
+      if (searchResults.length) {
+        this.selectResultItem(e, searchResults[this.state.selectedResultIndex]);
+      }
+      this.closeSearch(e);
+      e.preventDefault();
+    } else if (e.key === "Tab") {
+      this.closeSearch(e);
+      e.preventDefault();
+    }
+  }
+
+  renderResults() {
+    const { symbolSearchResults, selectedResultIndex } = this.state;
+    const { query, enabled } = this.props;
+    if (query == "" || !enabled || !symbolSearchResults.length) {
```

Comment:
identity equality here might be useful as well, or, if not important, we can `!query`